test/collections/componentSet.test.ts

```diff
@@ -725,6 +737,87 @@ describe('ComponentSet', () => {
           version: testApiVersionAsString,
         },
       });
+      expect(resolveSpy.called).to.be.false;
+      expect(getCurrentApiVersionStub.called).to.be.false;
+    });
+
+    it('should get an API version from sfdx-project.json', async () => {
+      const sourceApiVersion = '58.0';
+      const set = ComponentSet.fromSource({
+        fsPaths: ['.'],
+        registry: registryAccess,
+        tree: manifestFiles.TREE,
+      });
+      const resolveStub = $$.SANDBOX.stub(SfProject.prototype, 'resolveProjectConfig').resolves({ sourceApiVersion });
+      expect(await set.getObject()).to.deep.equal({
+        Package: {
+          types: [
+            {
+              name: registry.types.customobjecttranslation.name,
+              members: ['a'],
+            },
+            {
+              name: registry.types.staticresource.name,
+              members: ['b', 'c'],
+            },
+          ],
+          version: sourceApiVersion,
+        },
+      });
+      expect(resolveStub.calledOnce).to.be.true;
+      expect(getCurrentApiVersionStub.called).to.be.false;
+    });
+
+    it('should get an API version from env var', async () => {
+      const set = ComponentSet.fromSource({
+        fsPaths: ['.'],
+        registry: registryAccess,
+        tree: manifestFiles.TREE,
+      });
+      process.env.SF_ORG_API_VERSION = testApiVersionAsString;
+      expect(await set.getObject()).to.deep.equal({
+        Package: {
+          types: [
+            {
+              name: registry.types.customobjecttranslation.name,
+              members: ['a'],
+            },
+            {
+              name: registry.types.staticresource.name,
+              members: ['b', 'c'],
+            },
+          ],
+          version: testApiVersionAsString,
+        },
+      });
+      expect(getCurrentApiVersionStub.called).to.be.false;
+    });
+
+    it('should default to hardcoded API version 58.0 as a last resort', async () => {
+      getCurrentApiVersionStub.reset();
+      const causeErr = new Error('HTTP 404 - mdcoverage url is down');
+      getCurrentApiVersionStub.throws(causeErr);
+      const set = ComponentSet.fromSource({
+        fsPaths: ['.'],
+        registry: registryAccess,
+        tree: manifestFiles.TREE,
+      });
+      expect(await set.getObject()).to.deep.equal({
+        Package: {
+          types: [
+            {
+              name: registry.types.customobjecttranslation.name,
+              members: ['a'],
+            },
+            {
+              name: registry.types.staticresource.name,
+              members: ['b', 'c'],
+            },
+          ],
+          version: '58.0',
```

Comment:
why version 58.0 as default?

Reply:
I just chose a recent one.  It's better than throwing an error.  We really should never get to this point.  ComponentSets should have apiVersion and sourceApiVersion set by the consumer.

Reply:
At some point version 58.0 will be too old and things may not function as expected on this version. In that case does it make sense to throw an error vs using a default value that is old?

Reply:
The code will warn with the `missingApiVersion` message should it ever get to this point and use v58.0, so if an end user sees this message they can do things to get better results should it fail.  If we throw, they will definitely need to do something to get it to work.  Both Shane and I felt that doing what we could to not fail is better.

In practice, the checks for an API version should never get past the `sourceApiVersion` and `apiVersion` set on the `ComponentSet`.  If users of SDR are creating ComponentSets without setting a `sourceApiVersion` and/or `apiVersion` then they shouldn't be creating a manifest from it.